src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  public class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+    startFromIndex : Nat,
+  ) {
+
+    var expectedNextIndex_ : Nat = startFromIndex;
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+
+    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > 120_000_000_000; // 2 minutes
+
+    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), { #StreamClosed }> {
+      if (isStreamClosed()) {
+        return #err(#StreamClosed);
+      };
+      if (firstIndex != expectedNextIndex_) {
+        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(streamId, chunk[index], firstIndex + index);
+      };
+      expectedNextIndex_ += chunk.size();
+      #ok();
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let sender = StreamSender<Int>(
+  ///   123,
+  ///   10,
+  ///   10,
+  ///   func (item) = 1,
+  ///   anotherCanister.appendStream,
+  /// );
+  /// sender.next(1);
+  /// sender.next(2);
+  /// .....
+  /// sender.next(12);
+  /// await* sender.sendChunk(); // will send (123, [1..10], 0) to `anotherCanister`
+  /// await* sender.sendChunk(); // will send (123, [11..12], 10) to `anotherCanister`
+  /// await* sender.sendChunk(); // will do nothing, stream clean
+  public class StreamSender<T>(
+    streamId : Nat,
+    maxSize : ?Nat,
+    weightLimit : Nat,
+    weightFunc : (item : T) -> Nat,
+    callback : (streamId : Nat, items : [T], firstIndex : Nat) -> async R.Result<(), StreamError>,
+  ) {
+
+    let queue : QueueBuffer.QueueBuffer<T> = QueueBuffer.QueueBuffer<T>();
+    // a head of queue before submitting lately failed chunk. Used for error-handling
+    var lowestError : { #Inf; #Val : Nat } = #Inf;
+
+    public func fullAmount() : Nat = queue.fullSize();
+    public func queuedAmount() : Nat = queue.queueSize();
+    public func nextIndex() : Nat = queue.nextIndex();
+
+    public func get(index : Nat) : ?T = queue.get(index);
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+    var heartbeatTimer : Nat = 0;
+    heartbeatTimer := Timer.recurringTimer(
+      #seconds 30,
+      func() : async () {
+        // if last call was more than 20 seconds ago, send empty chunk
+        if ((Time.now() - lastChunkTimestamp) > 20_000_000_000) {
+          try {
+            switch (await callback(streamId, [], queue.headIndex())) {
+              case (#ok) lastChunkTimestamp := Time.now();
+              case (#err err) Timer.cancelTimer(heartbeatTimer);
+            };
+          } catch (err) {
+            // pass
+          };
+        };
+      },
+    );
```

Comment:
Let's not have a timer inside this class. Sending the empty batch should be part of `sendChunk`. If `sendChunk` has nothing to send then it looks at `lastChunkTimestamp` and, if that is too far back, sends an empty batch. 

`sendChunk` is already triggered from outside at regular intervals. We don't need another timer.

Reply:
removed